server/src/main/java/org/elasticsearch/common/Rounding.java

```diff
@@ -1085,4 +1101,54 @@ public static Rounding read(StreamInput in) throws IOException {
                 throw new ElasticsearchException("unknown rounding id [" + id + "]");
         }
     }
+
+    /**
+     * Attempt to build a {@link Prepared} implementation that relies on pre-calcuated
+     * "round down" points. If there would be more than {@code max} points then return
+     * the original implementation, otherwise return the new, faster implementation.
+     */
+    static Prepared maybeUseArray(Prepared orig, long minUtcMillis, long maxUtcMillis, int max) {
+        long[] values = new long[1];
+        long rounded = orig.round(minUtcMillis);
+        int i = 0;
+        values[i++] = rounded;
+        while ((rounded = orig.nextRoundingValue(rounded)) <= maxUtcMillis) {
+            if (i >= max) {
+                return orig;
+            }
+            assert values[i - 1] == orig.round(rounded - 1);
+            values = ArrayUtil.grow(values, i + 1);
+            values[i++]= rounded;
+        }
+        return new ArrayRounding(values, i, orig);
+    }
+
+    /**
+     * Implementation of {@link Prepared} using pre-calculated "round down" points.
+     */
+    private static class ArrayRounding implements Prepared {
+        private final long[] values;
+        private int max;
```

Comment:
make it final?

Reply:
:+1:. I have no idea why I didn't do that the first time around.